dev/service-registry/service-locator/src/main/scala/com/lightbend/lagom/gateway/NettyServiceGateway.scala

```diff
@@ -224,18 +224,32 @@ class NettyServiceGateway(coordinatedShutdown: CoordinatedShutdown, config: Serv
       }
       for (header <- request.headers.asScala) {
         // Ignore connection headers
-        if (!header.getKey.equalsIgnoreCase(HttpHeaders.Names.CONNECTION)) {
+        if (!header.getKey.equalsIgnoreCase(HttpHeaderNames.CONNECTION.toString)) {
           newRequest.headers.add(header.getKey, header.getValue)
         }
       }
-      // todo - add Forwarded and other headers
+      // todo - add Forwarded and other headers.
+      // UPDATE: current impl adds `X-Forwarded-Host instead of `Forwarded`. The `todo` message above still applies.
+      // see https://github.com/lagom/lagom/pull/1572
+      val newHostHeaders: HttpHeaders = rebuildHostHeaders(request, downstreamAddress)
+      newRequest.headers.remove(HttpHeaderNames.HOST).add(newHostHeaders)
 
       // Ensure the correct connection header is added to keep the connection alive
       HttpUtil.setKeepAlive(newRequest, true)
 
       newRequest
     }
 
+    private def rebuildHostHeaders(originalRequest: HttpRequest, downstreamAddress: URI): HttpHeaders = {
+      val originalHostValue = Option(originalRequest.headers().get(HttpHeaderNames.HOST))
+
+      val xForwardedHost: Option[(String, String)] = originalHostValue.map(hostValue => "X-Forwarded-Host" -> hostValue)
+      (Seq(
+        HttpHeaderNames.HOST -> s"${downstreamAddress.getHost}:${downstreamAddress.getPort}"
+      ) ++ xForwardedHost.toSeq)
+        .foldLeft[HttpHeaders](new DefaultHttpHeaders())((acc, header) => acc.add(header._1, header._2))
+    }
+
```

Comment:
I think that an imperative style here will make it more obvious, for example:

```scala
val headers = 
  new DefaultHttpHeaders()
    .add(HttpHeaderNames.HOST, s"${downstreamAddress.getHost}:${downstreamAddress.getPort}")
      
  if (originalRequest.headers().contains(HttpHeaderNames.HOST))
    headers.add("X-Forwarded-Host", originalRequest.headers().get(HttpHeaderNames.HOST))
  else
    headers
```

Reply:
more readable. thanks